### PR TITLE
Use officialnanoaod file

### PR DESCRIPTION
**Pull request by saswatinandan**

### CODE REVIEW

**Comment by ktht** [comment]
I'll remove my comments here shortly and add new ones because apparently github doesn't automatically display them inline when they're made for a certain commit that's part of the PR but before the PR was opened.

File: Framework/test/produceNtuple_cfg.py [hunk 1]
```diff
@@ -121,8 +121,8 @@
     branchName_jets_ak4 = cms.string('Jet'),
     branchName_jets_ak8_Hbb = cms.string('FatJet'),
     branchName_subjets_ak8_Hbb = cms.string('SubJet'),
-    branchName_jets_ak8_Wjj = cms.string('FatJetAK8LSLoose'),
-    branchName_subjets_ak8_Wjj = cms.string('SubJetAK8LSLoose'), 
+    branchName_jets_ak8_Wjj = cms.string('FatJet'),
```
Comment (ktht) [inline Framework/test/produceNtuple_cfg.py:128]: This change reminds me that at some point we should try lepton subject fractions (read from `FatJet_lsf3` branch) to figure out if boosted AK8 jets contain the selected lepton.

File: Objects/interface/RecoLepton.h [hunk 2]
```diff
@@ -172,12 +168,12 @@ class RecoLepton : public ChargedParticle
...
   Double_t dz_;                 ///< d_{z}, distance on the z axis w.r.t PV
   Double_t relIso_;             ///< relative mini-isolation
-  Double_t pfRelIso04All_;      ///< PF relative isolation dR=0.3, charged component
+  Double_t pfRelIso03All_;      ///< PF relative isolation dR=0.3, charged component
   Double_t miniRelIsoCharged_;  ///< relative charged mini-isolation
   Double_t miniRelIsoNeutral_;  ///< relative neutral mini-isolation (PU corrected)
   Double_t sip3d_;              ///< significance of IP
   Double_t mvaRawTTH_;          ///< raw output of lepton MVA of ttH multilepton analysis
-  Double_t jetPtRatio_;         ///< ratio of lepton pT to pT of nearby jet
+  Double_t jetRelIso_;         ///< ratio of lepton pT to pT of nearby jet
```
Comment (ktht) [inline Objects/interface/RecoLepton.h]: Please update the comment to:
```relative isolation in matched jet (1/ptRatio-1, pfRelIso04_all if no matched jet)```

File: Objects/src/RecoElectron.cc [hunk 3]
```diff
@@ -109,7 +101,7 @@ RecoElectron::etaSC() const
 Double_t
 RecoElectron::phiSC() const
```
Comment (ktht) [inline Objects/src/RecoElectron.cc]: Are we actually using this function anywhere? If not, then I advocate for removing it.

File: Objects/src/RecoLepton.cc [hunk 4]
```diff
@@ -10,47 +10,40 @@
 
 Double_t
 RecoLepton::get_assocJet_pt(Double_t reco_pt,
-                            Double_t jetPtRatio)
+                            Double_t jetRelIso)
 {
+  double jetPtRatio = 1. / jetRelIso - 1.;
```
Comment (ktht) [inline Objects/src/RecoLepton.cc]: This is not correct: `jetRelIso` is defined as `1 / jetPtRatio - 1` if the lepton comes from a jet, otherwise it's equal to pT isolation wrt to all PF candidates around the lepton cone of size dR = 0.4. Ignoring the latter, then `jetPtRatio` should be equal to `1 / (1 + jetRelIso)`.
I would skip the intermediate computation of `jetPtRatio` and would just return `0.9 * reco_pt * (jetRelIso + 1)` here, like shown [here](https://indico.cern.ch/event/812025/contributions/3475878/attachments/1867083/3070589/gp-fr-run2b.pdf) on slide 7.

File: Objects/src/RecoLepton.cc [hunk 5]
```diff
@@ -218,15 +211,15 @@ RecoLepton::mvaRawTTH_cut() const
...
-  return jetPtRatio_;
+  return jetRelIso_;
 }
 
 Double_t
-RecoLepton::jetRelIso() const
+RecoLepton::jetPtRatio() const
 {
-  return 1. / jetPtRatio_ - 1.;
+  return 1. / jetRelIso_ - 1.;
```
Comment (ktht) [inline Objects/src/RecoLepton.cc]: Please update to `1 / (1 + jetRelIso_)` as mentioned in my previous comment.

File: Readers/src/EventReader.cc [hunk 6]
```diff
@@ -125,7 +128,6 @@ EventReader::EventReader(const edm::ParameterSet& cfg)
 
   era_ = get_era(cfg.getParameter<std::string>("era"));
   isMC_ = cfg.getParameter<bool>("isMC");
-  readGenMatching_ = isMC_ && !cfg.getParameter<bool>("redoGenMatching");
```
Comment (ktht) [inline Readers/src/EventReader.cc:129]: How is this going to work when run on data?

File: Readers/src/GenJetReader.cc [hunk 7]
```diff
@@ -90,7 +90,6 @@ GenJetReader::setBranchAddresses(TTree * tree)
     bai.setBranchAddress(jet_eta_, branchName_eta_);
     bai.setBranchAddress(jet_phi_, branchName_phi_);
     bai.setBranchAddress(jet_mass_, branchName_mass_);
-    bai.setBranchAddress(jet_pdgId_, branchName_pdgId_);
```
Comment (ktht) [inline Readers/src/GenJetReader.cc:93]: Why remove?

File: Readers/src/LHEParticleReader.cc [hunk 8]
```diff
@@ -23,7 +23,7 @@ LHEParticleReader::LHEParticleReader(const edm::ParameterSet & cfg)
   , particle_pdgId_(nullptr)
   , particle_status_(nullptr)
 {
-  branchName_obj_ = cfg.getParameter<std::string>("branchName"); // default = "LHEPart"
+  branchName_obj_ = "LHEPart"; // default = "LHEPart"
```
Comment (ktht) [inline Readers/src/LHEParticleReader.cc]: Even though the branch names like these won't change, I prefer that we specify all these "magic strings" like "LHEPart" all in one place, and read them dynamically in the C++ code. In other words, please revert the change.

File: Readers/src/RecoElectronReader.cc [hunk 9]
```diff
@@ -148,7 +148,7 @@ RecoElectronReader::setBranchAddresses(TTree * tree)
     bai.setBranchAddress(sigmaEtaEta_, branchName_sigmaEtaEta_);
     bai.setBranchAddress(HoE_, branchName_HoE_);
     bai.setBranchAddress(deltaEta_, branchName_deltaEta_);
-    bai.setBranchAddress(deltaPhi_, branchName_deltaPhi_);
+    //    bai.setBranchAddress(deltaPhi_, branchName_deltaPhi_);
```
Comment (ktht) [inline Readers/src/RecoElectronReader.cc]: Remove the comment.

File: Readers/src/RecoElectronReader.cc [hunk 10]
```diff
@@ -216,19 +212,14 @@ RecoElectronReader::read() const
           gElectronReader->sigmaEtaEta_[idxLepton],
           gElectronReader->HoE_[idxLepton],
           gElectronReader->deltaEta_[idxLepton],
-          gElectronReader->deltaPhi_[idxLepton],
+            //          gElectronReader->deltaPhi_[idxLepton],
```
Comment (ktht) [inline Readers/src/RecoElectronReader.cc]: Remove the comment.

File: Readers/src/RecoLeptonReader.cc [hunk 11]
```diff
@@ -204,26 +204,18 @@ RecoLeptonReader::setBranchAddresses(TTree * tree)
...
-    bai.setBranchAddress(pfRelIso04_all_, branchName_pfRelIso04_all_);
+    bai.setBranchAddress(pfRelIso03_all_, branchName_pfRelIso03_all_);
     bai.setBranchAddress(relIso_all_, branchName_relIso_all_);
     bai.setBranchAddress(relIso_chg_, branchName_relIso_chg_);
-    bai.setBranchAddress(relIso_neu_, branchName_relIso_neu_);
     bai.setBranchAddress(sip3d_, branchName_sip3d_);
     bai.setBranchAddress(mvaRawTTH_, branchName_mvaRawTTH_);
-    bai.setBranchAddress(jetPtRatio_, branchName_jetPtRatio_);
-    bai.setBranchAddress(jetPtRel_, branchName_jetPtRel_, -1.);
-    for(const auto & kv: branchNames_jetBtagCSV_)
```
Comment (ktht) [inline Readers/src/RecoLeptonReader.cc:215]: Also remove the member variable from class definition.

File: Selectors/src/RecoElectronCollectionSelectorFakeable.cc [hunk 12]
```diff
@@ -32,7 +32,7 @@ RecoElectronSelectorFakeable::RecoElectronSelectorFakeable(Era era,
   , invert_nLostHits_(false)
   , min_nLostHits_fornLostHitsInversion_(1)
   , invert_conversionVeto_(false)
-  , useAssocJetBtag_(false)
+  , useAssocJetBtag_(true)
```
Comment (ktht) [inline Selectors/src/RecoElectronCollectionSelectorFakeable.cc]: Remove the member variable from all fakeable and tight classes, since it has no meaning anymore, as there is no ambiguity resolve anymore if there is just one kind of jet that is associated with a given lepton.

File: Readers/src/RecoJetReaderAK4.cc [hunk 13]
```diff
@@ -341,17 +336,12 @@ RecoJetReaderAK4::read() const
         qgl,
         gInstance->jet_bRegCorr_[idxJet],
         gInstance->jet_bRegRes_[idxJet],
-        gInstance->jet_pullEta_[idxJet],
-        gInstance->jet_pullPhi_[idxJet],
-        gInstance->jet_pullMag_[idxJet],
         jet_id,
         gInstance->jet_puId_[idxJet],
         gInstance->jet_genMatchIdx_[idxJet],
-        gInstance->jet_jetIdx_[idxJet],
```
Comment (ktht) [inline Readers/src/RecoJetReaderAK4.cc:350]: I think if you drop the jet position completely, then you cannot create the associations between reco leptons and reco jets. You probably need to store the position of the jet as they're read in, ie pass `idxJet` here. This way, when someone sorts the jets (because of eg JES or JER corrections which move the jets around a bit), then the association won't be lost.

**Review by ktht — CHANGES_REQUESTED** [review]
These are my first set of comments. I wasn't able to spot the piece of code that create the associations between reco jets and other reco objects explicitly. Has it implemented yet?

File: Readers/src/RecoJetReaderAK4.cc (hunk 13 above)
Comment (saswatinandan) [reply Readers/src/RecoJetReaderAK4.cc:350]: I discussed with Christian about this and he told that this jetIdx is used while cleaning the jets and currently we do cleaning by dr. If we need this we can add them in future but currently reduce the dependency of post-processing step

Comment (veelken) [reply Readers/src/RecoJetReaderAK4.cc:350]: Hi, 

I realized I may have made a mistake regarding this point. Karl, can you remind me in which analyzes we use jet cleaning by dR and in which we use cleaning by index ? 
I have a related question: do we compute the index that we use for the cleanining ourselves in the postprocessing or do we use the default nanoAOD implementation in
  https://github.com/HEP-KBFI/cmssw/blob/master/PhysicsTools/NanoAOD/python/nano_cff.py#L32-L38
I agree with Saswati that it is good to trim down as much as possible the postprocessing code, but if we can use the index computed by default in nanoAOD for the cleaning then I would vote for keeping the jet cleaning by index.

Comment (ktht) [reply Readers/src/RecoJetReaderAK4.cc:350]: We use the index-based cleaning in ttH and HH-multilepton analysis, and dR-based cleaning in HH-bbWW. The indices to the jet collection are determined during NanoAOD production. The only task from our part is to actually read the indices that are attributed to the selected reco leptons and taus, find the corresponding jets and remove the jets.

Index-based cleaning is less indiscriminate, since it cleans up to one jet per lepton, while dR-based cleaning can remove multiple jets per lepton. If it's not possible to keep both methods and switch between either if needed, then I propose we stick to the index-based approach. That's what other groups also seem to do, at least in ttH analysis. And unlike with HH-multilepton, we definitely need to collaborate and synchronize with other groups on the event selection.

File: Readers/src/GenJetReader.cc (hunk 7 above)
Comment (saswatinandan) [reply Readers/src/GenJetReader.cc:93]: This branch is not available in nanoaod file /hdfs/local/snandan/hhAnalysis/2017/official_nanoaod_file/F1B4499C-4A59-D142-8C2E-8364C1C6F90Cii.root but  present in previous file /hdfs/local/karl/ttHNtupleProduction/2017/2021Mar05_wPresel_nom_all_hh_multilepton/ntuples/signal_ggf_nonresonant_node_sm_hh_4v/0000/tree_1.root. I see this branch is available https://github.com/HEP-KBFI/TallinnNtupleProducer/blob/main/Readers/src/GenJetReader.cc#L79 . May be we need to turn on read_partonFlavour_, if this is the case please let me know and I'll do that

Comment (ktht) [reply Readers/src/GenJetReader.cc:93]: In the old software, `produceNtuple` used `GenJetReader` to read the branches but `GenParticleWriter` to write the collection. So, in order to read the collection back with `GenParticleReader` at the analysis level, the branch name had to be transcribed.

Since we don't have such problem anymore, I recommend that you read both partonFlavor and hadronFlavor of the gen jet, drop pdgId, and name the member variables accordingly.

File: Readers/src/RecoElectronReader.cc (hunk 10 above)
Comment (saswatinandan) [reply Readers/src/RecoElectronReader.cc]: done

File: Readers/src/RecoElectronReader.cc (hunk 9 above)
Comment (saswatinandan) [reply Readers/src/RecoElectronReader.cc]: done

File: Selectors/src/RecoElectronCollectionSelectorFakeable.cc (hunk 12 above)
Comment (saswatinandan) [reply Selectors/src/RecoElectronCollectionSelectorFakeable.cc]: done

File: Readers/src/LHEParticleReader.cc (hunk 8 above)
Comment (saswatinandan) [reply Readers/src/LHEParticleReader.cc]: done

File: Readers/src/EventReader.cc (hunk 6 above)
Comment (saswatinandan) [reply Readers/src/EventReader.cc:129]: corrected mistake

File: Objects/src/RecoLepton.cc (hunk 5 above)
Comment (saswatinandan) [reply Objects/src/RecoLepton.cc]: done

File: Objects/src/RecoLepton.cc (hunk 4 above)
Comment (saswatinandan) [reply Objects/src/RecoLepton.cc]: done

File: Objects/src/RecoElectron.cc (hunk 3 above)
Comment (saswatinandan) [reply Objects/src/RecoElectron.cc]: removed

File: Objects/interface/RecoLepton.h (hunk 2 above)
Comment (saswatinandan) [reply Objects/interface/RecoLepton.h]: done

File: Readers/src/RecoLeptonReader.cc (hunk 11 above)
Comment (saswatinandan) [reply Readers/src/RecoLeptonReader.cc:215]: done

File: Readers/src/GenJetReader.cc (hunk 7 above)
Comment (saswatinandan) [reply Readers/src/GenJetReader.cc:93]: done here https://github.com/HEP-KBFI/TallinnNtupleProducer/pull/3/commits/acab57b16864ce89c73f994eee3e70e6915ef846

File: EvtWeightTools/src/EvtWeightRecorder.cc [hunk 14]
```diff
@@ -776,6 +775,26 @@ EvtWeightRecorder::record_lheScaleWeight(const LHEInfoReader * const lheInfoRead
...
+void
+EvtWeightRecorder::record_gen_mHH_cosThetaStar(const LHEParticleCollection & lheParticles)
+{
+  std::vector<int> higgs;
+  TLorentzVector h1, h2, H;
+  for (unsigned int indx=0; indx<lheParticles.size(); indx++)
+  {
+    if ( lheParticles[indx].pdgId() ==25 ) higgs.push_back(indx);
+  }
+    if ( higgs.size() ==2 )
```
Comment (ktht) [inline EvtWeightTools/src/EvtWeightRecorder.cc]: This will only work if the ME is simulated up to Higgs bosons, which are then subsequently decayed by Pythia. In some HH signal samples (bbtautau I think) the Higgs bosons are already decayed at the ME level, which means that there are no status 1 LHE Higgs bosons in the collection, since NanoAOD stores only those by default. Previously, I got around this problem by storing status 2 LHE particles, because we had the luxury of running our own NanoAOD production, but this is not strictly necessary. One could also use the HH decay products to build the 4-momentum of both Higgses, which is why my proposal of storing LHE status 2 particles was rejected by XPOG.

So, using an if statement here would basically mean that computing mHH and cos(theta*) would fail silently for some HH signal samples. My proposal:
1. set `gen_mHH_` and `gen_cosThetaStar_` to some unphysical values in the constructor of this class, eg -1 and -2 would do;
2. throw an exception in `HHWeightInterfaceCouplings::getBinContent` if `TAxis::FindBin` hits an underflow bin (ie return 0). It basically means that whenever `HHWeightInterfaceLO::getWeight`  or `HHWeightInterfaceNLO::getWeight_LOtoNLO` are called with uninitialized mHH and cos(theta*), then it would result in an error;
3. protect the calls to `EvtWeightRecorder::record_gen_mHH_cosThetaStar` with an if statement in `produceNtuple` so that it'd be run only on HH signal samples. At the moment, the function is called for any process and the LHE particles are looped over for every single MC sample, which is simply suboptimal. If the function is called only for HH samples, then it would allow to extend the functionality of building 4-momenta of both Higgs bosons explicitly if they're already decayed at the ME level, and assert that the function really has found two Higgs bosons or their equivalent candidates.

By the way, I don't see `EvtWeightRecorder::record_hhWeight_*` called anywhere. How is this interface supposed to be used in the new FW?

File: Objects/src/GenParticle.cc [hunk 15]
```diff
@@ -15,6 +15,7 @@ GenParticle::GenParticle(const GenParticle & genParticle)
     , genParticle.status_
     , genParticle.statusFlags_
     , genParticle.genPartFlav_
+    , genParticle.genHadronFlav_
```
Comment (ktht) [inline Objects/src/GenParticle.cc]: Please revert the changes in `GenParticle` class. Instances of `GenJet` and `GenParticle` are objects of their own right. Just propagate `partonFlavor` of `GenJet` to `GenParticle`.

**Review by ktht — CHANGES_REQUESTED** [review]
Three items remain: ingredients to HH reweighting, reverting the changes in `GenPart`, and restoring the initialization of `RecoJetBaseIdx::idx_` in `RecoJetAK4` constructorn and its reader class (the cleaning functionality has already been implemented in the `EventReader` class).

File: Readers/src/RecoJetReaderAK4.cc (hunk 13 above)
Comment (saswatinandan) [reply Readers/src/RecoJetReaderAK4.cc:350]: I see some strange behavior. Though jetCleaningbyIndex is set True here
https://github.com/HEP-KBFI/TallinnNtupleProducer/blob/main/Framework/test/produceNtuple_cfg.py#L263 But in the code it is always considering this
https://github.com/HEP-KBFI/TallinnNtupleProducer/blob/main/Cleaners/interface/ParticleCollectionCleaner.h#L27 instead of this      https://github.com/HEP-KBFI/TallinnNtupleProducer/blob/main/Cleaners/interface/RecoJetCollectionCleanerByIndexAK4.h#L18

Comment (ktht) [reply Readers/src/RecoJetReaderAK4.cc:350]: According to the code that's currently on github, I don't see how this is possible. Please check again (grep for eg jetCleaningByIndex), recompile the code, and/or try adding debugging statements to make sure that the `EventReader` instance is correctly set up.

Comment (saswatinandan) [reply Readers/src/RecoJetReaderAK4.cc:350]: there is some problem in the code as in my current tree there is no branch Jet_jetIdx but still the code works fine. Even I removed all jetIdx variables from the code but still it works. But I'll add more debugging in the code and see later

Comment (ktht) [reply Readers/src/RecoJetReaderAK4.cc:350]: The `jetIdx` variables are available by default in NanoAOD, except for the jets themselves. This is because when we read the jet collection, the jets are already ordered as they're stored in NanoAOD. This is was not guaranteed in the old FW, where we had to rely on [`jetIdx` module](https://github.com/HEP-KBFI/tth-nanoAOD-tools/blob/master/python/postprocessing/modules/jetIdxProducer.py) that stored the original position of each jet in `Jet_jetIdx` branch.

Comment (saswatinandan) [reply Readers/src/RecoJetReaderAK4.cc:350]: I bring the jetIdx back here https://github.com/HEP-KBFI/TallinnNtupleProducer/commit/a4208400d6df3804106fb81f27bf0af61469d057 but I am still not able to find out why it is taking PartcleCollectionCleaner operator for AK4 jets. I took print out in main branch but it prints ParticleCollectionCleaner.May be I am doing some silly mistake somewhere

**Comment by veelken** [comment]
Hi Saswati,

there is a special piece of code that we use for cleaning AK4 jets by index:
  https://github.com/HEP-KBFI/TallinnNtupleProducer/blob/main/Cleaners/interface/RecoJetCollectionCleanerByIndexAK4.h

Tehnically, this class inherits from the class for general dR cleaning, which is defined here:
  https://github.com/HEP-KBFI/TallinnNtupleProducer/blob/main/Cleaners/interface/ParticleCollectionCleaner.h

We switch between cleaning AK4 jets by dR and cleaning by index using these lines of code:
  https://github.com/HEP-KBFI/TallinnNtupleProducer/blob/main/Readers/src/EventReader.cc#L186-L193

I hope that helps,

Christian

File: Readers/src/RecoJetReaderAK4.cc (hunk 13 above)
Comment (veelken) [reply Readers/src/RecoJetReaderAK4.cc:350]: Hi Saswati,

there is a special piece of code that we use for cleaning AK4 jets by index:
https://github.com/HEP-KBFI/TallinnNtupleProducer/blob/main/Cleaners/interface/RecoJetCollectionCleanerByIndexAK4.h

Tehnically, this class inherits from the class for general dR cleaning, which is defined here:
https://github.com/HEP-KBFI/TallinnNtupleProducer/blob/main/Cleaners/interface/ParticleCollectionCleaner.h

We switch between cleaning AK4 jets by dR and cleaning by index using these lines of code:
https://github.com/HEP-KBFI/TallinnNtupleProducer/blob/main/Readers/src/EventReader.cc#L186-L193

I hope that helps,

Christian

File: EvtWeightTools/src/EvtWeightRecorder.cc (hunk 14 above)
Comment (saswatinandan) [reply EvtWeightTools/src/EvtWeightRecorder.cc]: One thing is not clear to me, if Higgs are decyed at ME level how to build Higgs boson from the decay product using LHE info. I don't see any branch of mother in LHE branches

Comment (ktht) [reply EvtWeightTools/src/EvtWeightRecorder.cc]: You can build one Higgs candidate from a SFOS pair of particles that have an invariant mass of 125 GeV (+/- 1%).

File: Readers/src/RecoJetReaderAK4.cc (hunk 13 above)
Comment (saswatinandan) [reply Readers/src/RecoJetReaderAK4.cc:350]: Hi @veelken ,

 I understand this, I already looked into this before but didn't understand why it enters the operator [here](https://github.com/HEP-KBFI/TallinnNtupleProducer/blob/main/Cleaners/interface/ParticleCollectionCleaner.h#L27) when it is called for jetIndexMatching instead of [here](https://github.com/HEP-KBFI/TallinnNtupleProducer/blob/main/Cleaners/interface/RecoJetCollectionCleanerByIndexAK4.h#L18). So for cross check I requested Laurits to check and he also told me that it never enters the operator inside RecoJetCollectionCleanerByIndex. I request someone who are using main branch to check this. May be I made some changes in my branch which I can't indentify.

Comment (ktht) [reply Readers/src/RecoJetReaderAK4.cc:350]: So I had a look at the issue myself and I confirm that it does not work. The reason being, we define `jetCleanerAK4_dR04_` in terms of base class:
https://github.com/HEP-KBFI/TallinnNtupleProducer/blob/fa246726febeed601a0ca88bc4471bfd862b65cb/Readers/interface/EventReader.h#L144
And then we call the `operator()` function, which naturally points to the base class:
https://github.com/HEP-KBFI/TallinnNtupleProducer/blob/fa246726febeed601a0ca88bc4471bfd862b65cb/Readers/src/EventReader.cc#L563-L565
The natural solution here would be to promote the base class function to `virtual`, so that the ambiguity could be resolved at runtime, but C++ forbids declaring template functions as `virtual`. For a minimal working example, see https://ideone.com/OLqPyn and try changing the template function in the base class to `virtual`.

I don't have a good solution other than to go back to what I initially implemented in the old FW: that we have a dedicated class for `RecoJetCollectionCleanerByIndexAK4` that does not inherit from `ParticleCollectionCleaner`, create two separate instances of the cleaning classes, and resolve ambiguity between them with a simple if-else clause / ternary expression in `EventReader`.

By the way, I would also move the cleaning classes to stack instead of having them on heap. I don't see a clear benefit in declaring pointers and then having to call `->operator()(...)` explicitly.

Comment (veelken) [reply Readers/src/RecoJetReaderAK4.cc:350]: Hi,

I agree with Karl's suggestion. Saswati, please change 
  https://github.com/HEP-KBFI/TallinnNtupleProducer/blob/main/Cleaners/interface/RecoJetCollectionCleanerByIndexAK4.h#L8
to
  https://github.com/HEP-KBFI/tth-htt/blob/master/interface/RecoJetCollectionCleanerByIndex.h#L4
and then use lines like
  https://github.com/HEP-KBFI/tth-htt/blob/master/bin/analyze_2lss_1tau.cc#L1337-L1340
to switch between jet cleaning by dR and jet cleaning by index.

I also agree with Karl's comment that we can replace
  RecoJetCollectionCleanerAK4 * jetCleanerAK4_dR04_; // heap
by
  RecoJetCollectionCleanerAK4 jetCleanerAK4_dR04_; // stack
in https://github.com/HEP-KBFI/TallinnNtupleProducer/blob/fa246726febeed601a0ca88bc4471bfd862b65cb/Readers/interface/EventReader.h#L144-L145.
The reason why I used pointers is that I can declare a base-class pointer and assign this pointer either to the base-class (RecoJetCollectionCleanerAK4) or to the derived class (RecoJetCollectionCleanerByIndexAK4).

Comment (saswatinandan) [reply Readers/src/RecoJetReaderAK4.cc:350]: done [here](https://github.com/HEP-KBFI/TallinnNtupleProducer/pull/3/commits/d0bddf50ec96d8fb3edf7767a44bc6efd8d2d829) . I think we can replace selector pointer also by object, right?

Comment (ktht) [reply Readers/src/RecoJetReaderAK4.cc:350]: In principle, yes, but I would keep the changes as minimal as possible at this point because this PR is huge already.

File: EvtWeightTools/interface/EvtWeightRecorder.h [hunk 16]
```diff
@@ -219,13 +221,18 @@ class EvtWeightRecorder
   void
   record_prescale(double weight);
 
+  std::pair<std::vector<int>, std::vector<int>>
+  findHiggs(const LHEParticleCollection & lheParticles, std::vector<int> & exclude);
```
Comment (ktht) [inline EvtWeightTools/interface/EvtWeightRecorder.h:225]: I think that this function can be moved under `protected` and made `const` (ie add `const` before the semicolon) or `static` because the function does not touch any member variables.

File: EvtWeightTools/src/EvtWeightRecorder.cc [hunk 17]
```diff
@@ -776,6 +777,84 @@ EvtWeightRecorder::record_lheScaleWeight(const LHEInfoReader * const lheInfoRead
...
+  std::vector<int> higgs;
+  for (unsigned int indx1=0; indx1<lheParticles.size(); indx1++)
+  {
+    if ( lheParticles[indx1].pdgId() == 25 )
+    {
+      higgs.push_back(indx1);
+      if ( higgs.size() == 2 ) break;
+      continue;
+    }
+    if ( indx1 == lheParticles.size()-1 ) break;
```
Comment (ktht) [inline EvtWeightTools/src/EvtWeightRecorder.cc:794]: I think this line is not needed, because the for loop two lines below won't be executed if `indx1 == lheParticles.size()-1`.

**Review by ktht — APPROVED** [review]
Aside from two minor comments I think this PR is ready to be merged.

File: EvtWeightTools/src/EvtWeightRecorder.cc (hunk 17 above)
Comment (saswatinandan) [reply EvtWeightTools/src/EvtWeightRecorder.cc:794]: What about if higgs.size()==2 but indx1 !=lheParticles.size()-1 i.e there are other lhe particles which are not required anymore but still enter the loop until indx1 ==lheParticles.size()-1 is satisfied.
